src/Signal-Estimator-GUI/mainwindow.cpp

```diff
@@ -0,0 +1,304 @@
+#include "mainwindow.h"
+#include "ui_mainwindow.h"
+
+
+MainWindow::MainWindow(QWidget *parent) :
+    QMainWindow(parent),
+    ui(new Ui::MainWindow)
+{
+
+    if (!checkSignalEstimator()){ // open error window if signal-estimator is not found
+        SigEstNotFound w;
+        w.exec();
+        return;
+    }
+    ui->setupUi(this);
+
+    curve2->attach(this->ui->OutputSig); // output curve
+    curve1->attach(this->ui->OutputSig); // input curve
+    curve1->setPen(QColor(Qt::GlobalColor(Qt::blue))); // set input curve color to blue
+    curve2->setPen(QColor(Qt::GlobalColor(Qt::red))); // set input curve color to red
+
+    QVector<QString> in = getInputDevices();
+    QVector<QString> out = getOutputDevices();
+
+    this->ui->InputDevices->addItems(in.toList()); // add input devices to combobox
+    this->ui->OutputDevices->addItems(out.toList()); // add output devices to combobox
+
+    this->show();
+    return;
+}
+
+MainWindow::~MainWindow()
+{
+    delete ui;
+}
+
+void MainWindow::update_graphs(){ // update both input and output signal graphs
+    QVector<QPointF> in_current = in_data.getCurrentPoints();
+    // sort the data by x value so it doesn't jump randomly across graph
+    std::sort(in_current.begin(),in_current.end(),QPFcompare());
+    QVector<QPointF> out_current = out_data.getCurrentPoints();
+    std::sort(out_current.begin(),out_current.end(),QPFcompare());
+    //this->ui->ErrorLabel_2->setText(QString("updating graphs"));
+
+    this->curve1->setSamples(in_current);
+    this->curve2->setSamples(out_current);
+
+   // this->ui->InputSig->updateAxes();
+    //this->ui->InputSig->replot();
+
+    this->ui->OutputSig->updateAxes();
+    this->ui->OutputSig->replot();
+    QApplication::processEvents();
+
+    if (this->update_plots == true){ // restart timer for graphing
+        this->timer->start();
+    }
+}
+
+
+
+void MainWindow::read_graph_data(){
+     std::string buffer;
+
+     char* fullstr;
+     QByteArray arr;
+
+     this->proc->setReadChannel(QProcess::StandardOutput);
+     arr = this->proc->readLine(); // read line from proc
```

Comment:
1. We should check canReadLine() before calling readLine(). If canReadLine() returns false, than there is no complete line in QProcess yet, so we should return and wait until the signal is emitted next time.
2. It's possible that there are more than one line available for read. But the signal wont be re-emitted until *new* data arrives. So we should read lines in a loop until canReadLine() returns false.